setup.cfg

```diff
@@ -49,6 +49,9 @@ test =
 [options.packages.find]
 where = src
 
+[tool:pytest]
+norecursedirs = tests/integration/*
```

Comment:
If you'd download these into a tempdir on test run you would not need this 🤔

Reply:
But cloning to a tempdir is wastefull, unless you want to go for a semi-temp directory, which would avoid re-downloading between runs.

Reply:
It's only integration tests, so their not expected to be run at every invocation; so it's not that bad. Though they could be semi-temp. Alternatively, consider commiting the ``.tar.gz`` and extracting that into a temp folder. And yes you should be recreating the structure of these test projects at every run to ensure that build happens on a clean start, so you don't pass because of some left over files from a previous run.

Reply:
I don't think it's any more wasteful on CI than using submodules, since you do a clone of the submodules on each run anyway, but it is true that it might be annoying locally if you have no cache. I think a semi-temporary directory is fine as well.

Reply:
> I don't think it's any more wasteful on CI than using submodules, since you do a clone of the submodules on each run anyway, but it is true that it might be annoying locally if you have no cache. I think a semi-temporary directory is fine as well.

Note you'd need to implement clean functionality on semi-temporary directories, which adds code overhead. You don't need to worry about this if you e.g. just commit a .tar.gz and extract that on every run.

Reply:
> It's only integration tests, so their not expected to be run at every invocation; so it's not that bad. Though they could be semi-temp.

They aren't, but when they run, I want them to run fast. This is not a deal breaker but has 

> Alternatively, consider commiting the `.tar.gz` and extracting that into a temp folder.

Commit a `.tar.gz` to the repo? Is this what you mean?

> And yes you should be recreating the structure of these test projects at every run to ensure that build happens on a clean start, so you don't pass because of some left over files from a previous run.

Agreed.

> I don't think it's any more wasteful on CI than using submodules, since you do a clone of the submodules on each run anyway, but it is true that it might be annoying locally if you have no cache. I think a semi-temporary directory is fine as well.

It isn't, but it is wasteful on my machine. CI time is free, mine isn't. I don't want to download tarballs for all projects every time I want to run the integration tests, that will take a long time, and won't let me run the tests offline. Committing a tarball to the repo receives a big fat no from me, but I am not sure if this is what Bernát means. There are countless resources on the internet explaining why this is bad, but if anyone wants, I can expand on it. I guess we could use git lfs, but that is severely more PITA that submodules and Github doesn't give us much space. Which leaves us with the semi-temporary directory solution.

Personally I don't understand the issue of using submodules, those files are only needed by the developers, and we can have pytest detect if they are not present and print the clone command ready for you to copy+paste.

Anyway... it seems like having semi-temporary directory is an acceptable solution for everyone so lets go with it.

We can pull [`filelock`](https://github.com/benediktschmitt/py-filelock) as a test dependency and download the files to a semi-temporary directory. `filelock` is needed to handle test runs in parallel.

This semi-temporary directory could be cached by the CI to speed up things in the future.

Reply:
> Personally I don't understand the issue of using submodules

I'm suprised how you haven't been bitten/annoyed at submodules yet. But whenever I did use them quickly got to regret it. And feels to me @pganssle had the same experience.

> This semi-temporary directory could be cached by the CI to speed up things in the future.

Ok, but make sure to add some clean-feature so we can ensure that builds always start with clean semi-temporary folders. This is critical to ensure things work by design and not by leftovers from earlier runs.

Reply:
> I'm suprised how you haven't been bitten/annoyed at submodules yet. But whenever I did use them quickly got to regret it. And feels to me @pganssle had the same experience.

I have, but that is mitigated by having checks on the code.

> Ok, but make sure to add some clean-feature so we can ensure that builds always start with clean semi-temporary folders. This is critical to ensure things work by design and not by leftovers from earlier runs.

Of course, we would cache the semi-temporary directory with the tarballs, each tarball would be extracted to a temporary directory for the test run. We only keep the tarballs, so everything is definitely clean :grin:.

Reply:
> Ok, but make sure to add some clean-feature so we can ensure that builds always start with clean semi-temporary folders. This is critical to ensure things work by design and not by leftovers from earlier runs.

A cleaning feature should be easy enough, but I think we might be fine on this front anyway because nothing will be cached between CI runs, and end users can run `git clean` to clean all local directory caches.

Also, I think it makes sense to pin to an exact commit for each of these things to start with, for the purposes of reproducibility and stability. I don't think integration tests like this are a great fit for this current workflow, they're just considerably better than what we have now. Longer term I think the workflow should be revamped in such a way that we won't care as much about how stable the integration tests are, and hopefully by then we won't be worried about using any sort of semi-temporary directory or any strategy like that.

Reply:
The current tests are actually pinned to a commit. See https://github.com/FFY00/python-build/tree/tests-rework/tests/integration.

Reply:
Implemented the semi-temporary dir approach.